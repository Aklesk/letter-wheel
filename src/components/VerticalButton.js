import React from 'react';
import PropTypes from 'prop-types';
import { Image, Text, TouchableWithoutFeedback, View } from 'react-native';
import { observer } from 'mobx-react';
import { action, observable } from 'mobx';

import VerticalButtonStyle from '../styles/VerticalButtonStyle';

import verticalButtonYellowPassive from '../images/verticalButtonYellowPassive.png';
import verticalButtonYellowActive from '../images/verticalButtonYellowActive.png';

@observer
export default class Button extends React.Component {
  static propTypes = {
    content: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
    ]),
    onPress: PropTypes.func,
  }
  static defaultProps = {
    content: '',
    onPress() {},
  }

  @action beginPress = () => {
    this.isPressing = true;
  }
  @action endPress = () => {
    this.isPressing = false;
  }
  @observable isPressing = false;

  render() {
    const unpressed = verticalButtonYellowPassive;
    const pressed = verticalButtonYellowActive;

    return (
      <View style={VerticalButtonStyle.wrapper}>
        <Image
          fadeDuration={0}
          resizeMode="stretch"
          source={pressed}
          style={[
            VerticalButtonStyle.image,
            { opacity: this.isPressing ? 1 : 0 },
          ]}
        />
        <Image
          fadeDuration={0}
          resizeMode="stretch"
          source={unpressed}
          style={[
            VerticalButtonStyle.image,
            { opacity: this.isPressing ? 0 : 1 },
          ]}
        />
        <TouchableWithoutFeedback
          onPress={this.props.onPress}
          onPressIn={this.beginPress}
          onPressOut={this.endPress}
        >
          <View style={VerticalButtonStyle.contentWrapper}>
            <Text
              style={VerticalButtonStyle.text}
            >
              {this.props.content.toUpperCase()}
            </Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}
