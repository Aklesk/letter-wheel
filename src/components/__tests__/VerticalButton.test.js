import React from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import { mount } from 'enzyme';
import { isAction, isObservableProp, runInAction } from 'mobx';

import VerticalButton from '../VerticalButton';

describe('Button component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const render = mount((
      <VerticalButton />
    ));
    expect(render).toMatchSnapshot();
  });

  it('provides an \'isPressing\' observable', () => {
    const render = mount((
      <VerticalButton />
    ));
    expect(render.instance().isPressing).toEqual(false);
    expect(isObservableProp(render.instance(), 'isPressing')).toBeTruthy();
  });

  it('provides a \'beginPress\' action method', () => {
    const render = mount((
      <VerticalButton />
    ));
    expect(render.instance().beginPress).toEqual(expect.any(Function));
    expect(isAction(render.instance().beginPress)).toBeTruthy();
    expect(render.instance().isPressing).toEqual(false);
    expect(render.instance().beginPress()).toBeUndefined();
    expect(render.instance().isPressing).toEqual(true);
  });

  it('provides a \'endPress\' action method', () => {
    const render = mount((
      <VerticalButton />
    ));
    runInAction(() => { render.instance().isPressing = true; });
    expect(render.instance().endPress).toEqual(expect.any(Function));
    expect(isAction(render.instance().endPress)).toBeTruthy();
    expect(render.instance().isPressing).toEqual(true);
    expect(render.instance().endPress()).toBeUndefined();
    expect(render.instance().isPressing).toEqual(false);
  });

  it('renders a touchable that manages isPressing state and triggers onPress', () => {
    const onPress = jest.fn();
    const render = mount((
      <VerticalButton
        onPress={onPress}
      />
    ));
    const touchable = render.find(TouchableWithoutFeedback).at(0);
    expect(touchable.props().onPressIn).toEqual(render.instance().beginPress);
    expect(touchable.props().onPressOut).toEqual(render.instance().endPress);
    expect(touchable.props().onPress).toEqual(onPress);
  });
});
