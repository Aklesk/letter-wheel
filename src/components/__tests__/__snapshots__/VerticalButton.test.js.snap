// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches the snapshot 1`] = `
<Button
  content=""
  onPress={[Function]}
>
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <react-native-mock>
      <Image
        fadeDuration={0}
        resizeMode="stretch"
        source={
          Object {
            "testUri": "src/images/verticalButtonYellowActive.png",
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": undefined,
              "justifyContent": "center",
              "width": undefined,
            },
            Object {
              "opacity": 0,
            },
          ]
        }
      >
        <react-native-mock />
      </Image>
      <Image
        fadeDuration={0}
        resizeMode="stretch"
        source={
          Object {
            "testUri": "src/images/verticalButtonYellowPassive.png",
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": undefined,
              "justifyContent": "center",
              "width": undefined,
            },
            Object {
              "opacity": 1,
            },
          ]
        }
      >
        <react-native-mock />
      </Image>
      <TouchableWithoutFeedback
        onPress={[Function]}
        onPressIn={[Function]}
        onPressOut={[Function]}
      >
        <react-native-mock>
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingBottom": 16.941176470588236,
                "paddingLeft": 22.588235294117645,
                "paddingRight": 22.588235294117645,
                "paddingTop": 16.941176470588236,
              }
            }
          >
            <react-native-mock>
              <Text
                style={
                  Object {
                    "color": "#000000",
                    "fontFamily": "Grundschrift-Bold",
                    "fontSize": 22.588235294117645,
                  }
                }
              >
                <react-native-mock />
              </Text>
            </react-native-mock>
          </View>
        </react-native-mock>
      </TouchableWithoutFeedback>
    </react-native-mock>
  </View>
</Button>
`;
